kernel/host/xmicterm/host/xmicterm.cc

```diff
@@ -151,52 +147,54 @@ int main(int argc, char** argv)
     if (info_ptr->info == 0) sleep(1);
   } while (info_ptr->info == 0);
   std::cerr << "host info is at physical address "
-	    << (void*)info_ptr->info << std::endl;
+    << (void*)info_ptr->info << std::endl;
 
   auto info = micmem.access(PhysPtr<HostInfoTable>(info_ptr->info));
   cpu::clflush(info.addr());
   std::cerr << "debug stream is at physical address "
-	    << (void*)info->debugOut << std::endl;
+    << (void*)info->debugOut << std::endl;
 
   auto debugOutChannel = micmem.access(PhysPtr<HostInfoTable::DebugChannel>(info->debugOut));
   PCIeRingConsumer<HostInfoTable::DebugChannel> debugOut(debugOutChannel.addr());
 
-  undo_root(old_uid);
+  // drop root to allow file logger access to user specific directories
+  drop_root();
 
   // collect messages until whole message arrived
-  std::unordered_map<uint16_t, std::string> messages;
+  std::unordered_map<uint16_t, std::stringstream*> messages;
```

Comment:
Using string by value probably introduces a lot of heap allocations every time a string is copied or created. Stringstream cannot be used by value because copy constructor is deleted.
But reusing stringstream instead of removing them from the map should remove problems too.